Forge disk management (#73)
timescale/docs: timescale-forge/disk-management.md

1.  Detect if storage capacity begins to fill up
1.  Notify you about the growth
1.  Automatically activate overload protections
1.  Allow you to return your database to a normal state

This section explains what the various mechanisms are, and how to best make use
of them.

## Continuous storage monitoring

Timescale Forge continuously monitors the health and resource consumption of all
database services. You can check your health data by navigating to the `metrics`
tab in your service dashboard. These metrics are also monitored by the Timescale
operations team.

<img class="main-content__illustration" src="https://assets.timescale.com/images/diagrams/forge_metrics.png" alt="Timescale Forge metrics dashboard"/>

If your database exceeds a storage threshold of available resources, some
automated actions are triggered, including notifications, and preventative
actions.

## Automated user alerting

When your disk usage exceeds certain thresholds, you will recieve email
notifications. These notifications occur at:

*   75%
*   85%
*   95%

So that you aren't overwhelmed by automated message, the alerting thresholds use

> low- and high-watermarks, and only one message is sent per database per day.

Reviewer: ```suggestion
low- and high-watermarks, and only one message is sent per database per threshold per day.
```
Right now it is working like that. (Added changeset just to reflect facts - do not think that text should contain per/per/per... that makes it complicated.)

(my 2 cents:)
But I personally think that this can be changed in the future, and we can just say that number of messages is limited. 
cc @lukas-bernert

no downtime associated with this action. If your database is in read-only mode,
the read-only protection is automatically removed, and you can begin writing
data immediately.

<img class="main-content__illustration" src="https://assets.timescale.com/images/diagrams/forge_resources.png" alt="Timescale Forge change resources"/>

## Storage recovery

If you need to perform actions on your database to reduce your data usage, you
can turn off read-only mode. For example, you need read-write access if you want
to compress data, delete rows or tables, or drop old data using data retention
policies.

<highlight type="warning"> We highly recommend that you do not manually enable
read-write mode on a database that is over 99% capacity. You should consider
increasing the disk size before you do this. Alternatively, you can enable
read-write access on an individual session, while leaving the database in
read-only mode. </highlight>

### Procedure: Enabling read-write access on an individual session
1.  Connect to your database using `psql` and turn off read-only protection for the current session:
    ```sql
    SET default_transaction_read_only TO off;
    ```

> 1.  Turn on compression:

Reviewer: I think when database is in read-only it's better either to clean-up it, or to resize the volume.
Trying to enable compression at this point can lead to filling database to 100%. :smile_cat: 
cc @lukas-bernert

Reviewer 2: thanks for the note @itroot 

So it would make sense to structure it this way:

2./3. (no particular priority) Drop older Chunks OR increase disk size

4./5. (no particular priority) Add a retention policy OR turn on compression

CC: @Loquacity

# Disk management

Timescale Forge contains several mechanisms for managing disk space on your
services. There are four key tasks that Forge performs to handle disk space:

1.  Detect if storage capacity begins to fill up

> 1.  Notify you about the growth

Reviewer 3: ```suggestion
1.  Notify you about the growth of storage consumption
```

1.  Automatically activate overload protections
1.  Allow you to return your database to a normal state

This section explains what the various mechanisms are, and how to best make use
of them.

## Continuous storage monitoring

Timescale Forge continuously monitors the health and resource consumption of all
database services. You can check your health data by navigating to the `metrics`
tab in your service dashboard. These metrics are also monitored by the Timescale
operations team.

<img class="main-content__illustration" src="https://assets.timescale.com/images/diagrams/forge_metrics.png" alt="Timescale Forge metrics dashboard"/>

If your database exceeds a storage threshold of available resources, some
automated actions are triggered, including notifications, and preventative
actions.

## Automated user alerting

When your disk usage exceeds certain thresholds, you will recieve email
notifications. These notifications occur at:

*   75%
*   85%
*   95%

So that you aren't overwhelmed by automated message, the alerting thresholds use

> low- and high-watermarks, and we limit the number of messages we send you.

Reviewer 3: ```suggestion
low- and high-watermarks, and we limit the frequency of messages we send you
about a particular service.
```

## Automated overload protection

If your database continues to increase in size past these thresholds, automated
overload protection is activated when your disk becomes 99% full. When this
happens, your database is put into read-only mode, and you receive a
notification on email and the Timescale Forge console shows the changed status.

When this happens, you can still query your database, but you cannot add any new

> data to it. This is to ensure that your disk does not fill up to 100%, which
> could prevent you from crashing due to an out of memory (OOM) error.

Reviewer 3: ```suggestion
data to it. This is to ensure that your disk does not fill up to 100%, and
thus helps prevent you from crashing due to an out of storage errors.
```

Author: ```suggestion
data to it. This is to ensure that your disk does not fill up to 100%, and 
helps prevent you from crashing due to an out of memory (OOM) error.
```

Better?

With your database in read-only mode, you need to decide if you are going to
increase your storage capacity, or reduce the size of your database. When you
have done that, you can also add a retention policy, or turn on compression, to
avoid the problem occurring again in the future.

## Online storage resizing

You can increase your storage size in the Timescale Forge console.

<highlight type="warning">

> You can only resize your services once every six hours.

Reviewer 3: ```suggestion
You can only increase your service's storage once every six hours, and you cannot
currently decrease your storage size once set. 
```

</highlight>

### Procedure: Increasing service resources
1.  In the Timescale Forge console, navigate to `Services` and click the service
you want to adjust. Navigate to the `Operations` tab, and go to the `Resources`
section.
1.  Adjust the sliders for CPU and disk size as required. If you increase the

> disk size past a certain point, you may also need to increase the CPU size to
> handle the increased disk size.

Reviewer 3: ```suggestion
disk size past a certain point, we also recommend increasing the CPU size to
handle the increased disk size (although not required).
```

1.  Review the new sizes and costs in the panel on the right-hand side, and

> click `Restart and apply` when you are happy with the changes.

Reviewer 3: ```suggestion
click `Restart and apply` when you are happy with the changes.  Restarts are only
required when you increase your CPU size (which typically take a 20-60 seconds).
Storage increases happen fully online without any downtime.
```

Author: ```suggestion
click `Restart and apply` when you are happy with the changes. Restarts are only 
required when you increase your CPU size (which typically taked 20-60 seconds).
Storage increases happen fully online without any downtime.
```

1.  The resources take a few seconds to increase, and when the increase is

> complete, your database is immediately available on the new resources. There is
> no downtime associated with this action. If your database is in read-only mode,
> the read-only protection is automatically removed, and you can begin writing
> data immediately.

Reviewer 3: ```suggestion
complete, your database is immediately available on the new resources. If your
database is in read-only mode, the read-only protection is automatically removed,
and you can begin writing data immediately.
```

<img class="main-content__illustration" src="https://assets.timescale.com/images/diagrams/forge_resources.png" alt="Timescale Forge change resources"/>

## Storage recovery

If you need to perform actions on your database to reduce your data usage, you
can turn off read-only mode. For example, you need read-write access if you want
to compress data, delete rows or tables, or drop old data using data retention
policies.

<highlight type="warning"> We highly recommend that you do not manually enable

> read-write mode on a database that is over 99% capacity. You should consider

Reviewer 3: ```suggestion
read-write mode on your entire database that is over 99% capacity. You should consider
```

Author: ```suggestion
read-write mode on an entire database that is over 99% capacity. You should consider
```

services. There are four key tasks that Forge performs to handle disk space:

1.  Detect if storage capacity begins to fill up
1.  Notify you about the growth
1.  Automatically activate overload protections
1.  Allow you to return your database to a normal state

This section explains what the various mechanisms are, and how to best make use
of them.

## Continuous storage monitoring

Timescale Forge continuously monitors the health and resource consumption of all
database services. You can check your health data by navigating to the `metrics`
tab in your service dashboard. These metrics are also monitored by the Timescale
operations team.

*Insert screenshot here*

If your database exceeds a storage threshold of available resources, some
automated actions are triggered, including notifications, and preventative
actions.

## Automated user alerting

When your disk usage exceeds certain thresholds, you will recieve email notifications. These notifications occur at:

*   75%
*   85%
*   95%

So that you aren't overwhelmed by automated message, the alerting thresholds use low- and high-watermarks, and only one message is sent per database per day.

## Automated overload protection

> If your database continues to increase in size past these thresholds, automated overload protection is activated when your disk becomes 99% full. When this happens, your database is put into read-only mode, and you receive a notification on email and in the Timescale Forge console.

Reviewer 4: The Forge console won't give a notification, but the change in status is visible so maybe something like:
```suggestion
If your database continues to increase in size past these thresholds, automated overload protection is activated when your disk becomes 99% full. When this happens, your database is put into read-only mode, you receive a notification email and the changed status is visible in the Timescale Forge console.
```
or similar

Author: ```suggestion
If your database continues to increase in size past these thresholds, automated overload protection is activated when your disk becomes 99% full. When this happens, your database is put into read-only mode, and you receive a notification on email and the Timescale Forge console shows the changed status.
```
??

> When this happens, you can still query your database, but you cannot add any new data to it. This is to ensure that your disk does not fill up to 100%, which could prevent you from recovering the database.

Reviewer 4: "When this happens" => "In read-only mode" ? 
Also, maybe more explicit about the consequences of 100% => OOM crash in the DB?  Dunno, others are more knowledgeable about what exactly happens

Author: ```suggestion
When your database is in read-only mode, you can still query the database, but you cannot add any new data to it. This is to ensure that your disk does not fill up to 100%, which could prevent you from getting an out of memory (OOM) error.
```

I thought that failing to recover sounded scarier than OOM 😉

Reviewer 4: maybe "prevent it from crashing due to an out of memory (OOM) error." is scarier?  @itroot is probably best to answer what would be the best way to put it.

Author: ```suggestion
When this happens, you can still query your database, but you cannot add any new data to it. This is to ensure that your disk does not fill up to 100%, which could prevent you from crashing due to an out of memory (OOM) error.
```

I'm happy with that, if @itroot gives me a 👍

With your database in read-only mode, you need to decide if you are going to increase your storage capacity, or reduce the size of your database.

## Online storage resizing

You can increase your storage size in the Timescale Forge console.

<highlight type="warning">
You can only resize your services once every six hours.
</highlight>

### Procedure: Increasing service resources
1.  In the Timescale Forge console, navigate to `Services` and click the service you want to adjust. Navigate to the `Operations` tab, and go to the `Resources` section.

> 1.  Adjust the sliders for CPU and disk size as required. If you increase the disk size past a certain point, you will also need to increase the CPU size to handle the increased disk size.

Reviewer 4: ```suggestion
1.  Adjust the sliders for CPU and disk size as required. If you increase the disk size past a certain point, you may also need to increase the CPU size to handle the increased disk size.
```
leave it in their hands to decide?

1.  Review the new sizes and costs in the panel on the right-hand side, and click `Restart and apply` when you are happy with the changes.

> 1.  The resources take a few seconds to increase, and when the increase is complete, your database is immediately available on the new resources. If your database is in read-only mode, the read-only protection is automatically removed, and you can begin writing data immediately.

Reviewer 2: add that storage resize happens with zero downtime?

*Insert screenshot here*

## Storage recovery

If you need to perform actions on your database to reduce your data usage, you can turn off read-only mode. For example, you need read-write access if you want to compress data, delete rows or tables, or drop old data using data retention policies.

<highlight type="warning">

> We highly recommend that you do not manually enable read-write mode on a database that is over 99% capacity. You should consider increasing the disk size before you do this. Alternatively, you can enable read-write access on individual session, while leaving the database in read-only mode.

Reviewer 4: ```suggestion
We highly recommend that you do not manually enable read-write mode on a database that is over 99% capacity. You should consider increasing the disk size before you do this. Alternatively, you can enable read-write access on an individual session, while leaving the database in read-only mode.
```

</highlight>

### Procedure: Enabling read-write access on an individual session
1.  Connect to your database using `psql` and turn off read-only protection for the current session:
    ```sql
    SET default_transaction_read_only TO off;
    ```
1.  Turn on compression:
    ```sql

>     ALTER TABLE purchases SET (

Reviewer 2: Just a general thought:
Are placeholders more appropriate here in the docs?
Specific examples like "purchases" and "sku" make sense in the blog post but in the docs as well?

Author: Done, but not sure if I've chosen the best placeholder text. LMK.

      timescaledb.compress,
      timescaledb.compress_segmentby = 'sku'
    );

    SELECT add_compression_policy('purchases', interval '1 day');
    ```
1.  You can also create a data retention policy to only retain, for example, data
    for 90 days. This starts working immediately on old data:
    ```sql
    SELECT add_retention_policy('purchases', interval '90 days');
    ```

> As soon as the storage consumption drops below the threshold, the read-only protection is automatically removed, and can start writing data again.

Reviewer 4: ```suggestion
As soon as the storage consumption drops below the threshold, the read-only protection is automatically removed, and you can start writing data again.
```
or "...the database can ingest data again"?

Author: I prefer to keep the action with the reader, rather than the system.
